src/algorithm/proj.rs

```diff
@@ -27,6 +28,41 @@ impl Reproject for PointArray {
     }
 }
 
+macro_rules! iter_geo_impl {
+    ($type:ty, $builder_type:ty, $push_func:ident) => {
+        impl<O: OffsetSizeTrait> Reproject for $type {
+            fn reproject(&self, proj: &Proj) -> Result<Self> {
+                let mut output_array = <$builder_type>::new();
```

Comment:
Here this instantiates a new empty array; the downside of this is repeated allocations into the builder as we fill it. With https://github.com/geoarrow/geoarrow-rs/pull/313, I added APIs to get the existing buffer lengths, so that we can preallocate the _exact_ size of the output array necessary, at least for any algorithm that we know doesn't change the number of coordinates in each geometry.